### PR TITLE
Shenli/fix hibernate bug

### PR DESCRIPTION
**Pull request by shenli**
Fix this bug:
create table t1 (c double);
create table t2 (c double);
insert into t1 value (73);
insert into t2 value (930);
select c from t1 union select c from t2;
+----+
| c  |
+----+
| 73 |
| 0  |
+----+

### CODE REVIEW

File: plan/plans/info.go [hunk 1]
```diff
@@ -361,8 +361,14 @@ func (isp *InfoSchemaPlan) fetchColumns(schemas []*model.DBInfo) {
 				if decimal == types.UnspecifiedLength {
 					decimal = 0
 				}
-				dataType := types.TypeToStr(col.Tp, col.Charset == charset.CharsetBin)
-				columnType := fmt.Sprintf("%s(%d)", dataType, colLen)
+				columnType := types.TypeToStr(col.Tp, col.Charset == charset.CharsetBin)
+				if col.Decimal == types.UnspecifiedLength {
```
Comment (siddontang) [inline plan/plans/info.go]: I see other similar handing for type(m, n) in util/types and column, maybe we can unify them. 
Btw, you can see that float and double don't have (m, n) even they have length and decimal.

Comment (shenli) [inline plan/plans/info.go]: Yes.
And for data type without decimal, n will always be UnSpecifiedLength until the data is sent to mysql-client.

Comment (shenli) [inline plan/plans/info.go]: PTAL @siddontang

File: util/types/field_type.go [hunk 2]
```diff
@@ -52,45 +52,51 @@ func NewFieldType(tp byte) *FieldType {
...
 }
 
-// String joins the information of FieldType and
-// returns a string.
-func (ft *FieldType) String() string {
-	ts := FieldTypeToStr(ft.Tp, ft.Charset)
-	ans := []string{ts}
+// SimpleString only consider Tp/CharsetBin/Flen/Deimal.
+// This is used for showing column type in infoschema.
+func (ft *FieldType) SimpleString() string {
```
Comment (siddontang) [inline util/types/field_type.go]: SimpleString looks strange.

File: util/types/field_type.go [hunk 3]
```diff
@@ -52,45 +52,51 @@ func NewFieldType(tp byte) *FieldType {
...
+	ts := TypeToStr(ft.Tp, ft.Charset)
+	suffix := ""
 	switch ft.Tp {
 	case mysql.TypeEnum, mysql.TypeSet:
```
Comment (siddontang) [inline util/types/field_type.go]: no need to handle float and double ?

File: util/types/field_type.go (hunk 2 above)
Comment (shenli) [inline util/types/field_type.go]: Any suggestion?

File: util/types/field_type.go (hunk 3 above)
Comment (shenli) [inline util/types/field_type.go]: float/double's decimal will always be UnSpecifiedLength

Comment (siddontang) [inline util/types/field_type.go]: we can create table using float(3,1) format.

File: mysqldef/util.go [hunk 4]
```diff
@@ -29,8 +30,25 @@ func GetDefaultFieldLength(tp byte) int {
...
+	case TypeBit, TypeBlob:
+		return -1
 	default:
 		//TODO: add more types
+		return -1
+	}
+}
+
+// GetDefaultDecimal returns the default decimal length for column.
+func GetDefaultDecimal(tp byte) int {
```
Comment (qiuyesuifeng) [inline mysqldef/util.go]: How about GetDefaultFieldDecimal?

File: tidb-server/server/driver_tidb.go [hunk 5]
```diff
@@ -272,8 +272,16 @@ func convertColumnInfo(fld *field.ResultField) (ci *ColumnInfo) {
 	ci.Schema = fld.DBName
 	ci.Flag = uint16(fld.Flag)
 	ci.Charset = uint16(mysql.CharsetIDs[fld.Charset])
-	ci.ColumnLength = uint32(fld.Flen)
-	ci.Decimal = uint8(fld.Decimal)
+	if fld.Flen == -1 {
```
Comment (qiuyesuifeng) [inline tidb-server/server/driver_tidb.go]: Why not using  types.UnspecifiedLength？

File: util/types/field_type.go [hunk 6]
```diff
@@ -52,45 +52,58 @@ func NewFieldType(tp byte) *FieldType {
 	}
 }
 
-// String joins the information of FieldType and
-// returns a string.
-func (ft *FieldType) String() string {
-	ts := FieldTypeToStr(ft.Tp, ft.Charset)
-	ans := []string{ts}
+// CompactStr only consider Tp/CharsetBin/Flen/Deimal.
```
Comment (qiuyesuifeng) [inline util/types/field_type.go]: considers

File: util/types/field_type.go [hunk 7]
```diff
@@ -52,45 +52,58 @@ func NewFieldType(tp byte) *FieldType {
...
+func (ft *FieldType) CompactStr() string {
+	ts := TypeToStr(ft.Tp, ft.Charset)
+	suffix := ""
 	switch ft.Tp {
 	case mysql.TypeEnum, mysql.TypeSet:
 		// Format is ENUM ('e1', 'e2') or SET ('e1', 'e2')
-		ans = append(ans, fmt.Sprintf("('%s')", strings.Join(ft.Elems, "','")))
+		es := make([]string, 0, len(ft.Elems))
+		for _, e := range ft.Elems {
+			e = strings.Replace(e, "'", "''", -1)
```
Comment (qiuyesuifeng) [inline util/types/field_type.go]: A little confused why here replace ' with '' ?

Comment (siddontang) [inline util/types/field_type.go]: mysql too.

File: mysqldef/util.go (hunk 4 above)
Comment (shenli) [inline mysqldef/util.go]: ok

Comment (shenli) [inline mysqldef/util.go]: Decimal is clear enough.

File: tidb-server/server/driver_tidb.go (hunk 5 above)
Comment (shenli) [inline tidb-server/server/driver_tidb.go]: ok

File: util/types/field_type.go (hunk 7 above)
Comment (shenli) [inline util/types/field_type.go]: mysql> create table t (c set("'e1'", "'e2'"));
Query OK, 0 rows affected (0.01 sec)

mysql> desc t;
+-------+------------------------+------+-----+---------+-------+
| Field | Type                   | Null | Key | Default | Extra |
+-------+------------------------+------+-----+---------+-------+
| c     | set('''e1''','''e2''') | YES  |     | NULL    |       |
+-------+------------------------+------+-----+---------+-------+
1 row in set (0.00 sec)

**Comment by shenli** [comment]
PTAL @siddontang

File: column/column.go [hunk 8]
```diff
@@ -123,7 +123,7 @@ const defaultPrivileges string = "select,insert,update,references"
 func (c *Col) GetTypeDesc() string {
 	var buf bytes.Buffer
 
-	buf.WriteString(types.FieldTypeToStr(c.Tp, c.Charset))
+	buf.WriteString(types.TypeToStr(c.Tp, c.Charset))
 	switch c.Tp {
```
Comment (siddontang) [inline column/column.go]: use CompactStr instead

Comment (shenli) [inline column/column.go]: PTAL

**Pull request merged**